File: backend_java/src/main/java/ua/kostenko/tasks/app/service/JwtService.java
package ua.kostenko.tasks.app.service;

import io.jsonwebtoken.*;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;
import ua.kostenko.tasks.app.exception.InvalidJwtTokenException;

import javax.crypto.SecretKey;
import java.time.Instant;
import java.time.LocalDateTime;
import java.time.ZoneId;
import java.util.Date;

/**
 * A service class for handling operations related to JSON Web Tokens (JWT).
 * <p>
 * This class provides methods for generating, validating, and extracting claims from JWT tokens,
 * as well as checking their expiration status.
 * </p>
 */
@Slf4j
@Service
@RequiredArgsConstructor
public class JwtService {

    /**
     * The expiration time of JWT tokens in minutes.
     */
    private static final int JWT_EXP_MINUTES = 15;

    /**
     * The main secret key used for signing and verifying JWT tokens.
     */
    private final SecretKey jwtMainKey;

    /**
     * Extracts the claims from the provided JWT token.
     * <p>
     * This method parses the JWT token using the provided secret key. If the token is expired,
     * it returns the claims from the expired token. If the token is invalid, an exception is thrown.
     * </p>
     *
     * @param token the JWT token
     *
     * @return the claims contained in the token
     *
     * @throws RuntimeException if the token is invalid
     */
    public Claims extractClaims(String token) {
        try {
            JwtParser parser = Jwts.parser().verifyWith(jwtMainKey).build();
            Jws<Claims> signedClaims = parser.parseSignedClaims(token);
            log.debug("Successfully extracted claims from token.");
            return signedClaims.getPayload();
        } catch (ExpiredJwtException ex) {
            log.warn("JWT token is expired: {}", ex.getMessage());
            return ex.getClaims(); // Return claims for further processing
        } catch (Exception ex) {
            log.error("Invalid JWT token: {}", ex.getMessage());
            throw new InvalidJwtTokenException("Invalid JWT token: " + ex.getMessage());
        }
    }

    /**
     * Checks if the JWT token is expired based on its claims.
     *
     * @param claims the claims extracted from the JWT token
     *
     * @return {@code true} if the token is expired, {@code false} otherwise
     */
    public boolean isTokenExpired(Claims claims) {
        Date expirationDate = claims.getExpiration();
        if (expirationDate == null) {
            log.warn("Token expiration date is missing.");
            return true;
        }

        // Get the current date
        LocalDateTime localDateTime = LocalDateTime.now();
        Instant instant = localDateTime.atZone(ZoneId.systemDefault()).toInstant();
        Date currentDate = Date.from(instant);

        boolean isExpired = expirationDate.before(currentDate);
        log.debug("Token expiration check: expired = {}", isExpired);
        return isExpired;
    }

    /**
     * Validates the provided JWT token against the provided username.
     * <p>
     * This method checks that the token is not expired and the username in the token matches
     * the provided username.
     * </p>
     *
     * @param token    the JWT token
     * @param username the username to validate against the token
     *
     * @return {@code true} if the token is valid and the username matches, {@code false} otherwise
     */
    public boolean validateToken(String token, String username) {
        if (token == null || username == null) {
            log.warn("Token or username is null during validation.");
            return false;
        }

        try {
            Claims claims = extractClaims(token);
            String claimUsername = claims.getSubject();

            if (claimUsername == null) {
                log.warn("Username claim is null in the token.");
                return false;
            }

            boolean isUsernameValid = username.equals(claimUsername);
            boolean isTokenValid = !isTokenExpired(claims);

            log.debug("Token validation result: usernameValid = {}, tokenValid = {}", isUsernameValid, isTokenValid);
            return isUsernameValid && isTokenValid;
        } catch (RuntimeException ex) {
            log.warn("Failed to validate JWT token: {}", ex.getMessage());
            return false;
        }
    }

    /**
     * Generates a new JWT token for the specified username.
     * <p>
     * The token is signed using the secret key and has an expiration time of 15 minutes.
     * </p>
     *
     * @param username the username to include in the token's subject claim
     *
     * @return the generated JWT token as a compact string
     */
    public String generateJwtToken(String username) {
        // Get the current time and the expiration time
        LocalDateTime timeNow = LocalDateTime.now();
        LocalDateTime timeExp = timeNow.plusMinutes(JWT_EXP_MINUTES);
        Instant instantNow = timeNow.atZone(ZoneId.systemDefault()).toInstant();
        Instant instantExp = timeExp.atZone(ZoneId.systemDefault()).toInstant();
        Date dateNow = Date.from(instantNow);
        Date dateExp = Date.from(instantExp);

        // Generate the token
        String token =
                Jwts.builder().subject(username).issuedAt(dateNow).expiration(dateExp).signWith(jwtMainKey).compact();

        log.info("Generated JWT token for username: {}, valid until: {}", username, dateExp);
        return token;
    }
}

File: backend_go/internal/services/jwt_service.go
package services

import (
	"errors"
	"github.com/golang-jwt/jwt/v5"
	"github.com/rs/zerolog/log"
	"go_backend/internal/api"
	"time"
)

const JwtExpMinutes = 15 // JWT expiration time in minutes

// jwtService implements the JwtService interface and provides methods for managing JWT tokens.
type jwtService struct {
	jwtMainKey []byte // The main key used for signing JWT tokens.
}

// NewJwtService creates a new JwtService with the given main key.
//
// Parameters:
//   - jwtMainKey: A byte slice representing the key used for signing tokens.
//
// Returns:
//   - api.JwtService: A new instance of JwtService.
func NewJwtService(jwtMainKey []byte) api.JwtService {
	log.Debug().Msg("Creating new JwtService")
	return &jwtService{jwtMainKey: jwtMainKey}
}

// ExtractClaims extracts claims from a JWT token after verifying its signature.
//
// Parameters:
//   - token: A string representing the JWT token from which to extract claims.
//
// Returns:
//   - jwt.Claims: The extracted claims from the token.
//   - error: An error if the extraction fails.
func (s *jwtService) ExtractClaims(token string) (jwt.Claims, error) {
	log.Debug().Str("token", token).Msg("Extracting claims from token")

	claims := jwt.MapClaims{}
	parsedToken, err := jwt.ParseWithClaims(token, claims, func(token *jwt.Token) (interface{}, error) {
		return s.jwtMainKey, nil // Provide the signing key for verification
	}, jwt.WithoutClaimsValidation()) // Allow claims extraction without validating expiration

	if err != nil && !errors.Is(err, jwt.ErrTokenExpired) {
		log.Error().Err(err).Msg("Error parsing token")
		return nil, err
	}

	log.Debug().Msg("Claims extracted successfully")
	return parsedToken.Claims, nil
}

// IsTokenExpired checks if the JWT token has expired based on the claims.
//
// Parameters:
//   - claims: The claims extracted from the JWT token.
//
// Returns:
//   - bool: True if the token is expired, otherwise false.
func (s *jwtService) IsTokenExpired(claims jwt.Claims) bool {
	log.Debug().Msg("Checking if token is expired")

	expTime, err := claims.GetExpirationTime()
	if err != nil {
		log.Warn().Msg("Token expiration date is missing.")
		return true // Consider token expired if expiration time is missing
	}

	expirationTime := time.Unix(expTime.UnixNano(), 0)
	isExpired := expirationTime.Before(time.Now())
	log.Debug().Bool("isExpired", isExpired).Msg("Token expiration check result")
	return isExpired
}

// ValidateToken checks if the provided JWT token is valid and if the username matches the claims.
//
// Parameters:
//   - token: A string representing the JWT token to validate.
//   - username: A string representing the username to compare against the token claims.
//
// Returns:
//   - bool: True if the token is valid and the username matches, otherwise false.
func (s *jwtService) ValidateToken(token, username string) bool {
	log.Debug().Str("token", token).Str("username", username).Msg("Validating token")

	if token == "" || username == "" {
		log.Warn().Msg("Token or username is missing during validation.")
		return false
	}

	claims, err := s.ExtractClaims(token)
	if err != nil {
		log.Warn().Err(err).Msg("Failed to extract claims from token")
		return false
	}

	claimUsername, subjErr := claims.GetSubject()
	if subjErr != nil {
		log.Warn().Msg("Username claim is missing in the token.")
		return false
	}

	isUsernameValid := username == claimUsername
	isTokenValid := !s.IsTokenExpired(claims)

	log.Debug().Bool("isUsernameValid", isUsernameValid).Bool("isTokenValid", isTokenValid).Msg("Token validation result")
	return isUsernameValid && isTokenValid
}

// GenerateJwtToken generates a JWT token for the specified username.
//
// Parameters:
//   - username: A string representing the username for which to generate the token.
//
// Returns:
//   - string: The generated JWT token as a string.
//   - error: An error if token generation fails.
func (s *jwtService) GenerateJwtToken(username string) (string, error) {
	log.Debug().Str("username", username).Msg("Generating JWT token")

	now := time.Now()
	exp := now.Add(time.Minute * JwtExpMinutes) // Set token expiration time

	token := jwt.NewWithClaims(jwt.SigningMethodHS256, jwt.MapClaims{
		"sub": username,   // Subject of the token (the username)
		"iat": now.Unix(), // Issued at time
		"exp": exp.Unix(), // Expiration time
	})

	tokenString, err := token.SignedString(s.jwtMainKey)
	if err != nil {
		log.Error().Err(err).Msg("Error generating JWT token")
		return "", err
	}

	log.Debug().Str("username", username).Time("expiration", exp).Msg("Generated JWT token successfully")
	return tokenString, nil
}
